File: medcollab-backend/src/features/users/user.model.js
/**
 * USER MODEL
 *
 * Design decisions explained:
 *
 * 1. Phone as primary identity (not email)
 *    Indian doctors universally have a phone number.
 *    Email is inconsistent — many use college email that expires after graduation.
 *    Phone + OTP = zero password management.
 *
 * 2. No password field
 *    OTP-only auth. Nothing to hash, nothing to forget, nothing to breach.
 *
 * 3. fcmTokens is an array (not a string)
 *    A doctor might use the app on two phones (personal + hospital phone).
 *    We store all their device tokens and send notifications to all of them.
 *    Max 5 tokens per user — if they add a 6th, drop the oldest.
 *
 * 4. availability is embedded (not a separate collection)
 *    It changes frequently and is always read alongside the user.
 *    A separate collection would mean two DB calls to show "Dr. Priya — On Call".
 *
 * 5. isVerified vs isOnboarded
 *    isVerified: phone number confirmed via OTP
 *    isOnboarded: completed profile setup (name, role, speciality)
 *    We need both because a user might verify OTP but close the app before
 *    completing their profile. The Flutter app uses isOnboarded to decide
 *    whether to show the profile setup screen.
 *
 * 6. No patient data on the User model
 *    HIPAA-awareness: we never associate a doctor's account with specific patients.
 *    Patient references in handoffs use bed numbers and aliases only.
 */

const mongoose = require('mongoose');
const {
  USER_ROLES,
  AVAILABILITY_STATUS,
} = require('../../constants');

const availabilitySchema = new mongoose.Schema({
  status: {
    type: String,
    enum: Object.values(AVAILABILITY_STATUS),
    default: AVAILABILITY_STATUS.AVAILABLE,
  },
  // When "On Call" or "In OT", optionally show until when
  until: {
    type: Date,
    default: null,
  },
  // Human-readable note: "In CICU — emergency only"
  note: {
    type: String,
    maxlength: 100,
    default: '',
  },
  updatedAt: {
    type: Date,
    default: Date.now,
  },
}, { _id: false }); // No separate _id for embedded docs

const userSchema = new mongoose.Schema(
  {
    // ── Identity ──────────────────────────────────────────────────────────
    phone: {
      type: String,
      required: [true, 'Phone number is required'],
      unique: true,
      trim: true,
      // Store with country code: +919876543210
      // Validated at the service layer, not schema layer
    },

    name: {
      type: String,
      trim: true,
      maxlength: [100, 'Name cannot exceed 100 characters'],
    },

    // Optional: used for display in academic contexts
    displayTitle: {
      type: String,
      trim: true,
      maxlength: 20, // "Dr.", "Prof.", "DNB"
    },

    // ── Medical Identity ──────────────────────────────────────────────────
    role: {
      type: String,
      enum: Object.values(USER_ROLES),
      default: USER_ROLES.INTERN,
    },

    speciality: {
      type: String,
      trim: true,
      maxlength: 100,
      // e.g. "Cardiology", "General Medicine", "Orthopaedics"
    },

    // For PG residents — which year they are in
    pgYear: {
      type: Number,
      min: 1,
      max: 6,
    },

    // Medical college or hospital affiliation
    institution: {
      type: String,
      trim: true,
      maxlength: 200,
    },

    // City — useful for future community features
    city: {
      type: String,
      trim: true,
      maxlength: 100,
    },

    // ── Profile ───────────────────────────────────────────────────────────
    avatarUrl: {
      type: String,
      default: null,
    },

    bio: {
      type: String,
      maxlength: 300,
      default: '',
    },

    // ── Authentication ────────────────────────────────────────────────────
    isVerified: {
      type: Boolean,
      default: false, // true after first OTP verification
    },

    isOnboarded: {
      type: Boolean,
      default: false, // true after completing profile setup
    },

    // Soft deactivation — admin can disable an account without deleting it
    isActive: {
      type: Boolean,
      default: true,
    },

    // ── Push Notifications ────────────────────────────────────────────────
    // Array because doctors use multiple devices
    fcmTokens: {
      type: [String],
      default: [],
      validate: {
        validator: (tokens) => tokens.length <= 5,
        message: 'Maximum 5 device tokens allowed',
      },
    },

    // ── Availability ──────────────────────────────────────────────────────
    availability: {
      type: availabilitySchema,
      default: () => ({}),
    },

    // ── Preferences ──────────────────────────────────────────────────────
    // Store only settings that affect backend behaviour
    // UI preferences (theme, font size) stay on the device
    notifications: {
      emergencyAlerts: { type: Boolean, default: true },
      mentions: { type: Boolean, default: true },
      newMessages: { type: Boolean, default: true },
      handoffs: { type: Boolean, default: true },
      // Quiet hours: no notifications between these times
      quietHoursStart: { type: String, default: null }, // "22:00"
      quietHoursEnd: { type: String, default: null },   // "07:00"
    },

    // ── Tracking ─────────────────────────────────────────────────────────
    lastSeenAt: {
      type: Date,
      default: null,
    },
  },
  {
    timestamps: true, // Adds createdAt and updatedAt automatically
    toJSON: {
      transform: (doc, ret) => {
        // Never expose these fields in API responses
        delete ret.__v;
        return ret;
      },
    },
  }
);

// ── Indexes ─────────────────────────────────────────────────────────────────
// phone is already indexed via unique: true
// Index for searching members by institution (future: "find doctors at AIIMS")
userSchema.index({ institution: 1 });
userSchema.index({ role: 1, speciality: 1 });

// ── Instance Methods ─────────────────────────────────────────────────────────
// Methods available on a single user document

/**
 * Returns a safe public profile (no internal fields)
 * Used when sending user data to other users
 */
userSchema.methods.toPublicProfile = function () {
  return {
    _id: this._id,
    name: this.name,
    displayTitle: this.displayTitle,
    role: this.role,
    speciality: this.speciality,
    pgYear: this.pgYear,
    institution: this.institution,
    avatarUrl: this.avatarUrl,
    bio: this.bio,
    availability: this.availability,
    lastSeenAt: this.lastSeenAt,
    isOnboarded: this.isOnboarded,
    isVerified: this.isVerified,
  };
};

/**
 * Add an FCM token, removing duplicates and capping at 5
 */
userSchema.methods.addFcmToken = function (token) {
  // Remove duplicates
  this.fcmTokens = this.fcmTokens.filter((t) => t !== token);
  // Add to front (most recent first)
  this.fcmTokens.unshift(token);
  // Keep only 5
  if (this.fcmTokens.length > 5) {
    this.fcmTokens = this.fcmTokens.slice(0, 5);
  }
  return this.save();
};

const User = mongoose.model('User', userSchema);
module.exports = User;

File: medcollab-app/lib/features/auth/data/models/user_model.dart
import 'package:equatable/equatable.dart';
import 'package:medcollab_app/core/constants/app_enums.dart';
import 'package:medcollab_app/features/auth/data/models/availability_model.dart';
import 'package:medcollab_app/features/auth/data/models/notification_preferences_model.dart';

/// User profile — maps to backend `User.toPublicProfile()`.
///
/// Note: `isOnboarded` / `isVerified` / `phone` are not in `toPublicProfile()`
/// but may appear in future API versions. [isNewUser] from verify-otp drives
/// onboarding routing until then.
class UserModel extends Equatable {
  const UserModel({
    required this.id,
    this.phone,
    this.name,
    this.displayTitle,
    this.role = UserRole.intern,
    this.speciality,
    this.pgYear,
    this.institution,
    this.city,
    this.avatarUrl,
    this.bio = '',
    this.availability = const AvailabilityModel(),
    this.notifications = const NotificationPreferencesModel(),
    this.lastSeenAt,
    this.isVerified = false,
    this.isOnboarded = false,
  });

  factory UserModel.fromJson(Map<String, dynamic> json) {
    final id = json['_id'] ?? json['id'];
    return UserModel(
      id: id.toString(),
      phone: json['phone'] as String?,
      name: json['name'] as String?,
      displayTitle: json['displayTitle'] as String?,
      role: UserRole.fromString(json['role'] as String?),
      speciality: json['speciality'] as String?,
      pgYear: json['pgYear'] as int?,
      institution: json['institution'] as String?,
      city: json['city'] as String?,
      avatarUrl: json['avatarUrl'] as String?,
      bio: json['bio'] as String? ?? '',
      availability: json['availability'] is Map<String, dynamic>
          ? AvailabilityModel.fromJson(
              json['availability'] as Map<String, dynamic>,
            )
          : const AvailabilityModel(),
      notifications: json['notifications'] is Map<String, dynamic>
          ? NotificationPreferencesModel.fromJson(
              json['notifications'] as Map<String, dynamic>,
            )
          : const NotificationPreferencesModel(),
      lastSeenAt: json['lastSeenAt'] != null
          ? DateTime.tryParse(json['lastSeenAt'].toString())
          : null,
      isVerified: json['isVerified'] as bool? ?? false,
      isOnboarded: json['isOnboarded'] as bool? ?? false,
    );
  }

  final String id;
  final String? phone;
  final String? name;
  final String? displayTitle;
  final UserRole role;
  final String? speciality;
  final int? pgYear;
  final String? institution;
  final String? city;
  final String? avatarUrl;
  final String bio;
  final AvailabilityModel availability;
  final NotificationPreferencesModel notifications;
  final DateTime? lastSeenAt;
  final bool isVerified;
  final bool isOnboarded;

  String get displayName {
    if (name != null && name!.isNotEmpty) {
      final prefix = displayTitle != null && displayTitle!.isNotEmpty
          ? '$displayTitle '
          : '';
      return '$prefix$name';
    }
    return phone ?? 'Doctor';
  }

  /// True when the user has completed onboarding on the server.
  bool get hasMinimumProfile =>
      isOnboarded || (name != null && name!.trim().length >= 2);

  Map<String, dynamic> toJson() => {
        '_id': id,
        if (phone != null) 'phone': phone,
        if (name != null) 'name': name,
        if (displayTitle != null) 'displayTitle': displayTitle,
        'role': role.value,
        if (speciality != null) 'speciality': speciality,
        if (pgYear != null) 'pgYear': pgYear,
        if (institution != null) 'institution': institution,
        if (city != null) 'city': city,
        if (avatarUrl != null) 'avatarUrl': avatarUrl,
        'bio': bio,
        'availability': availability.toJson(),
        'notifications': notifications.toJson(),
        if (lastSeenAt != null) 'lastSeenAt': lastSeenAt!.toIso8601String(),
        'isVerified': isVerified,
        'isOnboarded': isOnboarded,
      };

  UserModel copyWith({
    String? id,
    String? phone,
    String? name,
    String? displayTitle,
    UserRole? role,
    String? speciality,
    int? pgYear,
    String? institution,
    String? city,
    String? avatarUrl,
    String? bio,
    AvailabilityModel? availability,
    NotificationPreferencesModel? notifications,
    DateTime? lastSeenAt,
    bool? isVerified,
    bool? isOnboarded,
  }) {
    return UserModel(
      id: id ?? this.id,
      phone: phone ?? this.phone,
      name: name ?? this.name,
      displayTitle: displayTitle ?? this.displayTitle,
      role: role ?? this.role,
      speciality: speciality ?? this.speciality,
      pgYear: pgYear ?? this.pgYear,
      institution: institution ?? this.institution,
      city: city ?? this.city,
      avatarUrl: avatarUrl ?? this.avatarUrl,
      bio: bio ?? this.bio,
      availability: availability ?? this.availability,
      notifications: notifications ?? this.notifications,
      lastSeenAt: lastSeenAt ?? this.lastSeenAt,
      isVerified: isVerified ?? this.isVerified,
      isOnboarded: isOnboarded ?? this.isOnboarded,
    );
  }

  @override
  List<Object?> get props => [
        id,
        phone,
        name,
        displayTitle,
        role,
        speciality,
        pgYear,
        institution,
        city,
        avatarUrl,
        bio,
        availability,
        notifications,
        lastSeenAt,
        isVerified,
        isOnboarded,
      ];
}
